### DISCUSSION TITLE
API throwing error: "Expected a list of items but got type \"dict\"."

### DISCUSSION THREAD

**Discussion by alexhoopes** [netbox-community/netbox#8879] [Q&A] [answered]
Trying to update device properties using /api/dcim/devices/, getting "API throwing error: "Expected a list of items but got type \"dict\"."

My JSON data:
```
{
   "name":"value",
   "device_type":"1",
   "device_role":"1",
   "site":"1",
   "status":"active",
   "face":"front",
   "custom_fields":{
      "field_1":"url",
      "sas_serial":"SERIAL-NUMBER"
   },
   "position":1,
   "platform":1,
   "serial":"SERIAL_2",
   "rack":1
}
```
Any guidance into where my issue is here?

**Comment by jcollie** [comment, marked as answer]
Is "/api/dcim/devices/" the complete URL that you are using? If so, that's the batch update API and works differently. I've never had a need for it so I can't talk much about how it's used. To update a specific device, you need to PUT/PATCH to "/api/dcim/devices/<id>/" where "<id>" is the ID number of the device to be updated.

**Reply by candlerb** [reply]
BTW, I think that reply is mangled by github markdown rendering, because unescaped text in angle brackets is interpreted as a HTML tags.

Use single backticks (\`) around a word or group of words to escape them, e.g. \`&lt;id&gt;\` renders as `<id>`.  Or you can use `&lt;` and `&gt;` entities instead of the brackets.  You can edit your existing post.

**Comment by aaron-newberry** [comment]
I am getting the same error when I try to update an interface on a device
https://netbox.my.com/api/dcim/interfaces/
{
"device": 21,
"module": 0,
"name": "Ethernet12",
"type": ["10gbase-t"],
"description": "this is a switch"
}

curl -X PATCH "https://netbox.my.com/api/dcim/interfaces/" -H "accept: application/json" -H "Content-Type: application/json" -H "X-CSRFToken: token blah" -d "{ "device": 21, "name": "Ethernet12", "type": ["1000base-t"], "description": "this is a switch "}"

**Reply by candlerb** [reply]
Send your PATCH to `/api/dcim/interfaces/NNN/` where NNN is the database ID of the interface in question.  If you don't know it, you'll have to search for it first.

Note that if you set "module" then it needs to point to a module ID (0 can't possibly be valid), and your interface type should be a single value, not a list.  Of course, if you don't want to change the interface name or description, then you can leave those fields out.

It *is* possible to send to `/api/dcim/interfaces/`, but in that case you are requesting a [bulk update](https://docs.netbox.dev/en/stable/integrations/rest-api/#updating-multiple-objects), which means that:
1. You must provide a JSON list as the outer value, containing zero or more objects
2. Each object must contain its ID

So it would become something like this:

```
[{
"id": NNN,
"name": "Ethernet12",
"type": "10gbase-t",
"description": "this is a switch port"
}]
```

However, you won't want to do this unless you have multiple interfaces you want to update simultaneously (and you know all their IDs)

**Reply by aaron-newberry** [reply]
Thank you for the information
